test/test_manifest.rb

```diff
@@ -684,6 +690,68 @@ module Autoproj
                 end
                 assert_equal "manifest #{File.join(ws.root_dir, 'test', 'manifest.xml')} of test from pkg_set lists 'dependency' as dependency, but it is neither a normal package nor an osdeps package. osdeps reports: cannot resolve dependency: dependency is not an osdep and it cannot be resolved as a source package", e.message
             end
+            it "parses an autoproj format manifest" do
```

Comment:
General comments about the tests:
- I know you copied the way the other tests were doing it (mocking everything), which is obviously horrible. Given that the rest of the file already does it, I can't ask you to change it, but I would nowadays go for actually writing files to disk instead.
- you don't test the ROS loader itself, only the conditional logic that would attempt to load `package.xml`

Reply:
> which is obviously horrible

The "horrible" part is because these tests are screaming for a refactor, there are lots of repeated code.

> I would nowadays go for actually writing files to disk instead

Generally speaking, I prefer using mocks in unit tests to make them more "unittary", so to speak. This is part of the problem of orogen's/orocos.rb's tests, they are way too dependent on external stuff, turning them into a major pain.

The integration part is left to integration tests.

> you don't test the ROS loader itself, only the conditional logic that would attempt to load package.xml

I deemed that unnecessary since `RosLoader´ is just a stripped down version of the other `Loader` but I can write those as well. No problem

Reply:
> Generally speaking, I prefer using mocks in unit tests to make them more "unittary", so to speak. This is part of the problem of orogen's/orocos.rb's tests, they are way too dependent on external stuff, turning them into a major pain.

I was there, and came back. If the code needs to touch the file system, then let it touch the filesystem. This actually makes the test less, not more, brittle as a new implementation of the same logic will have to deal with the filesystem the same way.

I don't see value in having "pure unitary" unit tests (I believe that there is mostly no such thing in the real world), and believe that the unit/integration test dichotomy is mostly false. And I'm not alone in this. Mocking your internal interfaces is actually pointless, as it makes the tests horrible to follow because of the amount of mocking, and extremely brittle when refactoring.

The problem with orogen is that to test a code generator, the best thing you can do is to actually generate and compile the code - which takes *really* long. This is the reason why I started to create the rock-build-tests stuff - let amake do the building and then check the result. Leaving orogen to do unit tests on the processing (i.e. before the codegen)

For orocos.rb, there was no easy way to create tasks and then test against these tasks, which meant that I had to actually compile code and run subprocesses (again). Now that there are ruby task contexts, I would *definitely* not attempt to "improve" the tests by mocking, but by simply creating ruby task contexts to test against, which will make the tests a lot faster and a lot more self-contained.

Mocking only gives you an illusion of "isolation" of your tests, it exposes the internal interfaces in ways that sometimes make little sense. The one *hugely* profitable use of mocking for me is to reduce the number of branches I have to test against - when a method "resolves" a complex inputs into a normalized (smaller) set of outputs that are used by the method, I happily mock that to only have to deal with the smaller normalized set instead of repeating tests against the bigger argument set.

Reply:
> it makes the tests horrible to follow because of the amount of mocking, and extremely brittle when refactoring.

This became more of a problem with these highly dynamic languages that allow mocking of entire class definitions (and pretty much anything you want). People stopped being disciplined about dependency injection and factories. The reason most tests have to be rewritten when the implementation changes is because we are mocking internals even though we are testing interfaces. If we were injecting the dependencies, we would only have to rewrite the tests when the dependencies' interfaces changed (which is actually a good thing in these cases). It also has the good side effect of making you really feel that you are making a possibly API breaking change.

Of course there are pros and cons on both sides (just like static vs. dynamic typing). I used to regret these dynamic language but I started to appreciate them since Ruby.. TypeScript was also a good surprise in that regard.

Reply:
> The reason most tests have to be rewritten when the implementation changes is because we are mocking internals even though we are testing interfaces. 

Agreed. Which is my original point. Test the interface ! It so happens that the filesystem is, for this class, part of the interface.